onnxruntime/core/framework/execution_frame.cc

```diff
@@ -35,33 +35,83 @@ IExecutionFrame::IExecutionFrame(const OrtValueNameIdxMap& ort_value_idx_map,
 
 IExecutionFrame::~IExecutionFrame() = default;
 
-// Return nullptr if index map to an value that is an unused optional input/output
-const OrtValue* IExecutionFrame::GetNodeInputOrOutputMLValue(int index) const {
-  int ort_value_idx = GetNodeIdxToMLValueIdx(index);
-  return ort_value_idx != NodeIndexInfo::kInvalidEntry ? &all_values_[ort_value_idx] : nullptr;
+#ifdef ENABLE_TRAINING
+void IExecutionFrame::UpdateFeeds(const std::vector<int>& feed_mlvalue_idxs, const std::vector<OrtValue>& feeds) {
+  for (size_t idx = 0, end = feed_mlvalue_idxs.size(); idx < end; ++idx) {
+    int ort_value_idx = feed_mlvalue_idxs[idx];
+    // we are sharing the underlying tensor/object for MLValue
+
+    ORT_ENFORCE(!all_values_[ort_value_idx].IsAllocated());
+
+    all_values_[ort_value_idx] = feeds[idx];
+  }
 }
 
-OrtValue* IExecutionFrame::GetMutableNodeInputOrOutputMLValue(int index) {
-  return const_cast<OrtValue*>(GetNodeInputOrOutputMLValue(index));
+void IExecutionFrame::UpdateFetches(const std::vector<int>& fetch_mlvalue_idxs, const std::vector<OrtValue>& fetches, const std::unordered_map<int, OrtValue>& initializers) {
+  if (!fetches.empty()) {
+    fetch_mlvalue_idxs_ = fetch_mlvalue_idxs;
+
+    auto num_fetches = fetch_mlvalue_idxs_.size();
+
+    for (size_t idx = 0; idx < num_fetches; ++idx) {
+      int ort_value_idx = fetch_mlvalue_idxs_[idx];
+
+      ORT_ENFORCE(!all_values_[ort_value_idx].IsAllocated());
+
+      all_values_[ort_value_idx] = fetches[idx];
+
+      // Copy the initializer if it is a fetch entry.
```

Comment:
do we have real scenario for fetching initializer? 
all our weights are fed as graph inputs, the remaining initializers are constants. 
I am trying to think of case where we need to fetch initializer.

Reply:
Its a legal corner case. The cost is minimal, just line 63-64 which is constant time operation.